File: web/src/config.test.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { afterEach, describe, expect, it, vi } from 'vitest';
import { API_BASE_URL } from './config';

describe('config API base url', () => {
  afterEach(() => {
    vi.unstubAllEnvs();
    vi.resetModules();
  });

  it('defaults to a relative /api path when no env override is set', () => {
    expect(API_BASE_URL).toBe('/api');
  });

  it('never resolves to an absolute localhost address by default', () => {
    expect(API_BASE_URL.startsWith('http://localhost')).toBe(false);
    expect(API_BASE_URL.startsWith('https://localhost')).toBe(false);
    expect(API_BASE_URL.startsWith('//localhost')).toBe(false);
  });

  it('strips a single trailing slash', () => {
    expect(API_BASE_URL.endsWith('/')).toBe(false);
  });

  it('uses an env override and strips its trailing slash', async () => {
    vi.stubEnv('VITE_API_BASE_URL', 'https://gateway.example/api/');
    vi.resetModules();

    const mod = await import('./config');

    expect(mod.API_BASE_URL).toBe('https://gateway.example/api');
  });
});

File: rmqctl/internal/config/config_test.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package config

import (
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

func TestSave(t *testing.T) {
	t.Run("round trip preserves config", func(t *testing.T) {
		directory := t.TempDir()
		if err := os.Chmod(directory, 0o700); err != nil {
			t.Fatal(err)
		}
		path := filepath.Join(directory, "config.yaml")
		want := Config{
			CurrentContext: "prod",
			Contexts: map[string]Context{
				"prod": {
					Server:  "https://studio.example.com",
					Cluster: "instance-prod",
					Credential: CredentialRef{
						AccessKeyRef: "env:RMQ_PROD_AK",
						SecretKeyRef: "env:RMQ_PROD_SK",
					},
				},
			},
		}
		store := NewStore()
		if err := store.Save(path, want); err != nil {
			t.Fatal(err)
		}
		got, err := store.Load(path)
		if err != nil {
			t.Fatal(err)
		}
		if !reflect.DeepEqual(got, want) {
			t.Fatalf("round trip mismatch:\ngot  %#v\nwant %#v", got, want)
		}
	})
}

func TestResolveCredential(t *testing.T) {
	t.Run("resolves env references with trim", func(t *testing.T) {
		values := map[string]string{"RMQ_AK": "admin-ak", "RMQ_SK": " secret with spaces "}
		credential, err := ResolveCredential(CredentialRef{
			AccessKeyRef: "env:RMQ_AK",
			SecretKeyRef: "env:RMQ_SK",
		}, func(name string) string { return values[name] })
		if err != nil {
			t.Fatal(err)
		}
		if credential.AccessKey != "admin-ak" || credential.SecretKey != "secret with spaces" {
			t.Fatalf("credential = %#v", credential)
		}
	})

}
